chipsec/hal/igd.py

```diff
@@ -0,0 +1,201 @@
+        if self.logger.HAL: self.logger.log( '[igd] Graphics MMIO and GTT (GTTMMADR): 0x%016X' % base )
+        return base
+
+    def get_GGTT_base(self):
+        gtt_off = 0x200000 if self.is_legacy_gen() else 0x800000
+        return self.get_GTTMMADR() + gtt_off
+
+    def get_PTE_size(self):
+        return 4 if self.is_legacy_gen() else 8
+
+    def read_GGTT_PTE(self, pte_num):
+        gtt_base = self.get_GGTT_base()
+        reg_off = (self.get_PTE_size()*pte_num)
+       
+        pte_lo = self.cs.mmio.read_MMIO_reg( gtt_base, reg_off )
+        pte_hi = 0
+        if self.get_PTE_size() == 8:
+            pte_hi = self.cs.mmio.read_MMIO_reg( gtt_base, reg_off + 4)
+        return (pte_lo | (pte_hi << 32))
+
+    def write_GGTT_PTE(self, pte_num, pte):
+        gtt_base = self.get_GGTT_base()
+        self.cs.mmio.write_MMIO_reg( gtt_base, self.get_PTE_size()*pte_num, pte & 0xFFFFFFFF)
+        if self.get_PTE_size() == 8:
+            self.cs.mmio.write_MMIO_reg( gtt_base, self.get_PTE_size()*pte_num + 4, pte >> 32)
+        return pte
+
+    def write_GGTT_PTE_from_PA(self, pte_num, pa):
+        pte = self.get_GGTT_PTE_from_PA( pa )
+        gtt_base = self.get_GGTT_base()
+        self.cs.mmio.write_MMIO_reg( gtt_base, self.get_PTE_size()*pte_num, pte & 0xFFFFFFFF )
+        if self.get_PTE_size() == 8:
+            self.cs.mmio.write_MMIO_reg( gtt_base, self.get_PTE_size()*pte_num + 4, pte >> 32)
+        return pte
+
+    def dump_GGTT_PTEs(self, num):
+        gtt_base = self.get_GGTT_base()
+        self.logger.log( '[igd] Global GTT contents:' )
+        ptes = self.cs.mmio.read_MMIO( gtt_base, num*self.get_PTE_size() )
+        pte_num = 0
+        for pte in ptes:
+            self.logger.log( 'PTE[%03d]: %08X' % (pte_num, pte) )
+            pte_num = pte_num + 1
+
+    def get_GGTT_PTE_from_PA(self, pa):
+        if self.is_legacy_gen():
+            return self.get_GGTT_PTE_from_PA_legacy( pa )
+        else:
+            return self.get_GGTT_PTE_from_PA_gen8( pa )
+
+    def get_GGTT_PTE_from_PA_legacy(self, pa):
+        #
+        # GTT PTE format:
+        # 0     - valid
+        # 2:1   - cache type (00 - reserved, 01 - UC, 10 - LLC only, 11 - MLC/LLC)
+        # 3     - GFDT
+        # 11:4  - PA bits 39:32
+        # 31:12 - PA bits 31:12
+        #
+        return ((pa & 0xFFFFF000) | ((pa>>32 & 0xFF) << 4) | 0x3)
+
+    def get_PA_from_PTE_legacy(self, pte):
+        return (((pte & 0x00000FF0) << 28) | (pte & 0xFFFFF000))
+
+    def get_GGTT_PTE_from_PA_gen8(self, pa):
+        return ((pa & ~0xFFF) | 0x1)
+
+    def get_PA_from_PTE_gen8(self, pte):
+        return (pa & ~0xFFF)
+
+    def get_PA_from_PTE(self, pte):
+        if self.is_legacy_gen():
+            return self.get_PA_from_PTE_legacy()
+        else:
+            return self.get_PA_from_PTE_gen8( pte )
+
+
+    def gfx_aperture_dma_read_write(self, address, size=0x4, value=None, pte_num=0):

```

Comment:
Would probably be better to swap arguments 'value' and 'pte_num' so that a caller could do read through arbitrary PTE without specifying named 'pte_num' argument.